Firestore/Swift/Source/Codable/FirestoreDecoder.swift

```diff
@@ -0,0 +1,1055 @@
+            
+            return value
+        }
+    }
+}
+
+@available(swift 4.0.0)
+fileprivate class _FirestoreDecoder : Decoder {
+    /// Options set on the top-level encoder to pass down the decoding hierarchy.
+    
+    // MARK: Properties
+    /// The decoder's storage.
+    fileprivate var storage: _FirestoreDecodingStorage
+    
+    /// The path to the current point in encoding.
+    fileprivate(set) public var codingPath: [CodingKey]
+    
+    /// Contextual user-provided information for use during encoding.
+    public var userInfo: [CodingUserInfoKey : Any] = [:]
+    
+    // MARK: - Initialization
+    /// Initializes `self` with the given top-level container and options.
+    init(referencing container: Any, at codingPath: [CodingKey] = []) {
+        self.storage = _FirestoreDecodingStorage()
+        self.storage.push(container: container)
+        self.codingPath = codingPath
+    }
+    
+    // MARK: - Decoder Methods
+    public func container<Key>(keyedBy type: Key.Type) throws -> KeyedDecodingContainer<Key> {
+        guard !(self.storage.topContainer is NSNull) else {
+            throw DecodingError.valueNotFound(KeyedDecodingContainer<Key>.self,
+                                              DecodingError.Context(codingPath: self.codingPath,
+                                                                    debugDescription: "Cannot get keyed decoding container -- found null value instead."))
+        }
+        
+        guard let topContainer = self.storage.topContainer as? [String : Any] else {
+            let context = DecodingError.Context(codingPath: codingPath, debugDescription: "Not a dictionary")
+            throw DecodingError.typeMismatch([String: Any].self, context)
+        }
+        
+        let container = _FirestoreKeyedDecodingContainer<Key>(referencing: self, wrapping: topContainer)
+        return KeyedDecodingContainer(container)
+    }
+    
+    public func unkeyedContainer() throws -> UnkeyedDecodingContainer {
+        guard !(self.storage.topContainer is NSNull) else {
+            throw DecodingError.valueNotFound(UnkeyedDecodingContainer.self,
+                                              DecodingError.Context(codingPath: self.codingPath,
+                                                                    debugDescription: "Cannot get unkeyed decoding container -- found null value instead."))
+        }
+        
+        guard let topContainer = self.storage.topContainer as? [Any] else {
+            let context = DecodingError.Context(codingPath: codingPath, debugDescription: "Not an array")
+            throw DecodingError.typeMismatch([Any].self, context)
+        }
+        
+        return _FirestoreUnkeyedDecodingContainer(referencing: self, wrapping: topContainer)
+    }
+    
+    public func singleValueContainer() throws -> SingleValueDecodingContainer {
+        return self
+    }
+}
+
+@available(swift 4.0.0)
+fileprivate struct _FirestoreDecodingStorage {
```

Comment:
Could this be named `DocumentDecodingStorage`?

My basic concern is that "Firestore decoding" is too general. For example, we could, in the future decode whole query snapshots, in which case a "Firestore decoder" is insufficient.